source/common/config/grpc_mux_impl.cc

```diff
@@ -210,6 +220,73 @@ void GrpcMuxImpl::onEstablishmentFailure() {
+    // should create a new entry for config_source_control_plane pointing to this grpc_service.
+    auto service_control_plane_info =
+        control_plane_config->service_control_plane_info().find(xds_service_);
+    if (service_control_plane_info == control_plane_config->service_control_plane_info().end()) {
+      // If service it self does not exist, this is the first time we are seeing it, so add it.
+      ::envoy::admin::v2alpha::ControlPlaneConfigDump_ServiceControlPlaneInfo
+          new_service_control_plane_info;
+      auto* new_config_source_control_info =
+          new_service_control_plane_info.add_config_source_control_plane();
+      populateControlPlaneConfigSource(control_plane_config.get(), new_config_source_control_info,
+                                       message.control_plane().identifier());
+      (*control_plane_config->mutable_service_control_plane_info())[xds_service_].MergeFrom(
+          new_service_control_plane_info);
+    } else {
+      // If service exists, check if the config source exists, if it is there update it else create
+      // a new config source.
+      if (!config_tracked_) {
+        // create a new config source with grpc_service and identifier.
+        auto* new_config_source_control_info =
+            (*control_plane_config->mutable_service_control_plane_info())[xds_service_]
+                .add_config_source_control_plane();
+        populateControlPlaneConfigSource(control_plane_config.get(), new_config_source_control_info,
+                                         message.control_plane().identifier());
+      } else {
+        // update the control plane identifier of existing grpc_service.
+        auto* existing_config_source_control_info =
+            (*control_plane_config->mutable_service_control_plane_info())[xds_service_]
+                .mutable_config_source_control_plane(config_source_index_);
+        populateControlPlaneConfigSource(control_plane_config.get(),
+                                         existing_config_source_control_info,
+                                         message.control_plane().identifier());
+      }
+    }
+    config_tracker_.addOrUpdateControlPlaneConfig(
+        xds_service_,
+        std::make_shared<envoy::admin::v2alpha::ControlPlaneConfigDump>(*control_plane_config));
+  }
+}
+
+void GrpcMuxImpl::populateControlPlaneConfigSource(
+    const envoy::admin::v2alpha::ControlPlaneConfigDump* control_plane_config,
+    envoy::admin::v2alpha::ControlPlaneConfigDump_ConfigSourceControlPlaneInfo* config_source_info,
+    const std::string& identifier) {
+  if (!config_tracked_) {
+    config_source_info->mutable_grpc_service()->MergeFrom(grpc_service_);
+    config_tracked_ = true;
+    auto service_control_plane_info =
+        control_plane_config->service_control_plane_info().find(xds_service_);
+    if (service_control_plane_info != control_plane_config->service_control_plane_info().end()) {
+      config_source_index_ =
```

Comment:
How does this work if entries in `config_source_control_plane()` are removed (and then maybe added back later)? Do you rely on some invariant that this is always monotonically increasing? How does that work if a server decides to stop using EDS in a later CDS update?

Reply:
That is a good question. I was thinkg about this and wondering about this optimization after pushed the commit - I think to be safe it is better to find the `index` every time. Will change it accordingly.